### PR TITLE
Add project setup with kubebuilder

### PR DESCRIPTION
**Pull request by Rakshith-R**
Kubebuilder requires no directories and pre-existing files present when creating a new project. 

Therefore, created a branch with kuberbuilder project and merged it with main branch. 

Getting this in, will make follow up pr's smaller and easier to review.

### CODE REVIEW

**Comment by Madhu-1** [comment]
@Rakshith-R looks  like PR is having conflict can you please check

**Comment by Rakshith-R** [comment]
> @Rakshith-R looks like PR is having conflict can you please check

its fixed now, please take a look

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,27 @@
+# Build the manager binary
+FROM golang:1.16 as builder
```
Comment (Madhu-1) [inline Dockerfile:2]: use 1.17 as it's available?

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+# Build
+RUN CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go build -a -o manager main.go
+
+# Use distroless as minimal base image to package the manager binary
+# Refer to https://github.com/GoogleContainerTools/distroless for more details
+FROM gcr.io/distroless/static:nonroot
+WORKDIR /
+COPY --from=builder /workspace/manager .
+USER 65532:65532
```
Comment (Madhu-1) [inline Dockerfile:25]: need to remove user and use the latest image or else we will have GRPC problem see https://github.com/csi-addons/volume-replication-operator/pull/49

File: Makefile [hunk 3]
```diff
@@ -0,0 +1,130 @@
+
+# Image URL to use all building/pushing image targets
+IMG ?= controller:latest
+# ENVTEST_K8S_VERSION refers to the version of kubebuilder assets to be downloaded by envtest binary.
+ENVTEST_K8S_VERSION = 1.22
```
Comment (Madhu-1) [inline Makefile:5]: use 1.23 as its latest?

File: Makefile [hunk 4]
```diff
@@ -0,0 +1,130 @@
+
+# Image URL to use all building/pushing image targets
+IMG ?= controller:latest
```
Comment (Madhu-1) [inline Makefile:3]: the image name need to be changed to the proper name?

**Review by Madhu-1 — COMMENTED** [review]
having a basic CI setup is needed as a follow-up PR.

File: Dockerfile (hunk 2 above)
Comment (Rakshith-R) [reply Dockerfile:25]: > need to remove user and use the latest image or else we will have GRPC problem see [csi-addons/volume-replication-operator#49](https://github.com/csi-addons/volume-replication-operator/pull/49)

controller will communicate with sidecar over network. 
For sidecar, we'll have a new build file with `gcr.io/distroless/static:latest` and no user.

Comment (Madhu-1) [reply Dockerfile:25]: are we going to have 2 dockerfiles? I thought we are going to have 1 docker image for both controller and sidecar.

Comment (Madhu-1) [reply Dockerfile:25]: Have you already verified that with the non-root image you are able to communicate over the wire?

File: Dockerfile (hunk 1 above)
Comment (Rakshith-R) [reply Dockerfile:2]: > use 1.17 as it's available?

This is the default version kubebuilder creates, I'll update the go version here and upgrade the packages in a follow-up pr.

File: Makefile (hunk 4 above)
Comment (Rakshith-R) [reply Makefile]: changed

File: Makefile (hunk 3 above)
Comment (Rakshith-R) [reply Makefile]: done

File: Dockerfile (hunk 1 above)
Comment (Rakshith-R) [reply Dockerfile]: changed

File: Dockerfile (hunk 2 above)
Comment (Rakshith-R) [reply Dockerfile:25]: > are we going to have 2 dockerfiles? I thought we are going to have 1 docker image for both controller and sidecar.

2 separate images, I'll let @nixpanic confirm. 

> Have you already verified that with the non-root image you are able to communicate over the wire?

root image is required for unix socket communication. 
Controller communicating with k8s api server is over the wire / tcp .
And it being default should work.

Comment (nixpanic) [reply Dockerfile:25]: yes, two container-images. Both contain a different executable, with only the functionality (and permissions) that they need. There is little overlap in functionality between controller and side-car, so it is cleaner to provide images for both functionalities. This also allows security checkers to validate the permissions that an image/executable needs.

File: Makefile [hunk 5]
```diff
@@ -1,6 +1,6 @@
 
 # Image URL to use all building/pushing image targets
-IMG ?= controller:latest
+IMG ?= quay.io/csi-addons/controller:latest
```
Comment (nixpanic) [inline Makefile:3]: Maybe this should include `kubernetes` or `k8s` in the name? The controller is Kubernetes specific and will not work on other container platforms.

**Comment by Rakshith-R** [comment]
I have created https://github.com/csi-addons/kubernetes-csi-addons/issues/9 to track task related to image name and build ci, deploy images.

(changing image name in Makefile needs additional changes in default dev environment.)

File: config/default/kustomization.yaml [hunk 6]
```diff
@@ -0,0 +1,74 @@
+# Adds namespace to all resources.
+namespace: kubernetes-csi-addons-system
+
+# Value of this field is prepended to the
+# names of all resources, e.g. a deployment named
+# "wordpress" becomes "alices-wordpress".
+# Note that it should also match with the prefix (text before '-') of the namespace
+# field above.
+namePrefix: kubernetes-csi-addons-
```
[the remark this answers is not in the document]
Comment (Rakshith-R) [reply config/default/kustomization.yaml:9]: > is the trailing `-` required?

seems to be, it'll be appended to all resources created.

> This repository is Kubernetes specific. So are deployments. Maybe drop the `kubernetes` prefix?

dropped

File: config/default/manager_auth_proxy_patch.yaml [hunk 7]
```diff
@@ -0,0 +1,27 @@
+# This patch inject a sidecar container which is a HTTP proxy for the
+# controller manager, it performs RBAC authorization against the Kubernetes API using SubjectAccessReviews.
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: controller-manager
```
[the remark this answers is not in the document]
Comment (Rakshith-R) [reply config/default/manager_auth_proxy_patch.yaml:6]: > rename?

This is only used in development environment and may not be used in actual deployment.

File: Dockerfile [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+# and so that source changes don't invalidate our downloaded layer
+RUN go mod download
+
+# Copy the go source
+COPY main.go main.go
+COPY api/ api/
+COPY controllers/ controllers/
+
+# Build
+RUN CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go build -a -o manager main.go
```
[the remark this answers is not in the document]
Comment (Madhu-1) [reply Dockerfile:18]: https://github.com/csi-addons/kubernetes-csi-addons/issues/11 for supporting multi-arch image. if we do that `arch`  will be input for dockerfile

**Review by Madhu-1 — APPROVED** [review]

**Comment by Rakshith-R** [comment]
rebased on top of https://github.com/csi-addons/kubernetes-csi-addons/pull/15. PTAL

**Review by Madhu-1 — APPROVED** [review]
LGTM. lets bring the enhancements as a followup PR's

**Comment by Madhu-1** [comment]
@nixpanic PTAL. if you don't have any comments we can merge this one and have follow-up PR's for enhancements.

**Review by yati1998 — COMMENTED** [review]
LGTM

**Review by nixpanic — APPROVED** [review]

**Pull request merged**